tensorrt_llm/_torch/pyexecutor/decoder.py

```diff
@@ -240,153 +225,162 @@ def _handle_stop_criteria(self, request: LlmRequest, new_token: int,
+        for request in decoder_state.scheduled_requests.context_requests:
+            assert not request.py_return_context_logits, "return_context_logits not implemented for TorchDecoder"
+
             if request.get_context_remaining_length() != 0:
-                idx += 1
+                logits_idx += 1
+                request_idx += 1
                 continue
 
             if request.state != LlmRequestState.GENERATION_COMPLETE:
-                new_token = new_tokens_list[idx]
+                new_token = new_tokens_list[logits_idx]
                 num_tokens = request.add_new_token(new_token, beam_idx)
                 self._handle_stop_criteria(request, new_token, num_tokens,
                                            beam_idx)
-            idx += 1
 
-        if hasattr(scheduled_requests, 'chunked_requests'):
-            for request in scheduled_requests.chunked_requests:
-                idx += 1
+            logits_idx += 1
+            request_idx += 1
 
-        extend_requests = []
-        generation_requests = []
-        for request in scheduled_requests.generation_requests:
-            if hasattr(
-                    request,
-                    'py_draft_tokens') and request.py_draft_tokens is not None:
-                extend_requests.append(request)
-            else:
-                generation_requests.append(request)
-
-        for request in extend_requests:
-            num_accepted = 0
-            if request.state != LlmRequestState.GENERATION_COMPLETE:
-                new_token = new_tokens_list[idx]
-                num_tokens = request.add_new_token(new_token, beam_idx)
-                self._handle_stop_criteria(request, new_token, num_tokens,
-                                           beam_idx)
+        if hasattr(decoder_state.scheduled_requests, 'chunked_requests'):
+            logits_idx += len(decoder_state.scheduled_requests.chunked_requests)
+
+        # sorted() is stable.
+        # TODO: can we ensure extend request (has_draft_tokens = True) are always before generation requests,
+        #  therefore this sorted() call can be avoided?
+        generation_requests = sorted(
+            decoder_state.scheduled_requests.generation_requests,
+            key=lambda request: hasattr(request, 'py_draft_tokens') and request.
+            py_draft_tokens is not None,
+            reverse=True)
+
+        for request in generation_requests:
+            if request.state == LlmRequestState.GENERATION_COMPLETE:
+                continue
 
-                # Accept draft tokens (if we have any) if and only if they match the new
-                # token exactly.
+            accepted_tokens, total_tokens = 1, 1
+            new_token = new_tokens_list[logits_idx]
+            num_tokens = request.add_new_token(new_token, beam_idx)
+            self._handle_stop_criteria(request, new_token, num_tokens, beam_idx)
+
+            if request.py_draft_tokens is not None:
```

Comment:
`request.py_draft_tokens is None` is called twice, one time in `sorted`, one time here. In the original implementation, it's only called one time. 

I suggest we keep the original implemention. For duplicate codes in `extend_requests` and `generation_requests`, we can write a helper function:

```python
    def _handle_request(request):     
        if request.state != LlmRequestState.GENERATION_COMPLETE:
                new_token = new_tokens_list[idx]
                num_tokens = request.add_new_token(new_token, beam_idx)
                self._handle_stop_criteria(request, new_token, num_tokens,
                                           beam_idx)
```

Reply:
I reverted most of the changes to `update_requests()`. They are not very meaningful without logits handling code.

Now the only changes are simplify skipping `chunked_requests` and remove `hasattr(request, "py_draft_tokens")` check. `py_draft_tokens` is always set by `LlmRequest.__init__()` now.